controller/eks-cluster-config-handler.go

```diff
@@ -973,30 +1009,36 @@ func (h *Handler) updateUpstreamClusterState(upstreamSpec *eksv1.EKSClusterConfi
 				return config, err
 			}
 		}
-		err := createNodeGroup(config, ng, eksService, svc)
+		err := createNodeGroup(config, ng, eksService, ec2Service, svc)
 		if err != nil {
 			return config, err
 		}
 		updatingNodegroups = true
 	}
 
 	// check if node groups need to be deleted
+	templateVersionsToDelete := make([]*string, 0)
 	for _, ng := range upstreamSpec.NodeGroups {
 		if _, ok := ngs[aws.StringValue(ng.NodegroupName)]; ok {
 			continue
 		}
-		_, err := eksService.DeleteNodegroup(
-			&eks.DeleteNodegroupInput{
-				ClusterName:   aws.String(config.Spec.DisplayName),
-				NodegroupName: ng.NodegroupName,
-			})
+		templateVersionToDelete, _, err := deleteNodeGroup(config, ng, eksService)
 		if err != nil {
 			return config, err
 		}
 		updatingNodegroups = true
+		if templateVersionToDelete != nil {
+			templateVersionsToDelete = append(templateVersionsToDelete, templateVersionToDelete)
+		}
 	}
 
 	if updatingNodegroups {
+		if len(templateVersionsToDelete) != 0 {
+			config = config.DeepCopy()
+			config.Status.Phase = eksConfigUpdatingPhase
+			config.Status.TemplateVersionsToDelete = templateVersionsToDelete
+			return h.eksCC.UpdateStatus(config)
```

Comment:
I updated `h.checkAndUpdate` so that it is not possible for this function (`h.updateUpstreamClusterState`) to be called with `Status.TemplateVersionsToDelete` being non-nil.